aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,127 @@
+#include <ATen/BatchedFallback.h>
+#include <ATen/VmapTransforms.h>
+
+namespace at {
+
+// Given a linear index, return the actual index.
+// Example: Given linear_idx = 3, sizes = [5, 2], we would return [1, 0]
+static SmallVector<indexing::TensorIndex,kVmapStaticDimVecSize>
+computeIndex(int64_t linear_idx, IntArrayRef sizes) {
+  SmallVector<indexing::TensorIndex,kVmapStaticDimVecSize> result;
+  result.reserve(sizes.size());
+  for (auto it = sizes.rbegin(); it != sizes.rend(); it++) {
+    auto remainder = linear_idx % *it;
+    result.push_back(remainder);
+    linear_idx -= remainder;
+    linear_idx /= *it;
+  }
+  std::reverse(std::begin(result), std::end(result));
+  return result;
+}
+
+void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+  auto num_returns = op.schema().returns().size();
+  TORCH_CHECK(!schema.is_mutable() && !schema.hasAnyAliasInfo(),
+              "Batching rule not implemented for ", schema, "; ",
+              "the fallback path doesn't work on in-place or view ops.");
+  TORCH_WARN("Batching rule not implemented for ", op.schema(), " falling back "
+             "to slow (for loop and stack) implementation");
+  TORCH_CHECK(std::all_of(op.schema().returns().begin(),
+                          op.schema().returns().end(),
+                          [] (const Argument& arg) { return arg.type() == TensorType::get(); }),
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We could not generate a fallback.");
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
```

Comment:
Is this morally OK? `torch::jit::Stack` is typedef'ed to be `vector<ivalue>`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/stack.h#L11 and I am using it as a vector throughout this function.

There are a lot of convenience functions for `torch::jit::Stack` (https://github.com/pytorch/pytorch/blob/b85df3709a9e92f88d2e86b4860571010bc06e62/aten/src/ATen/core/stack.h#L29-L49) but they're harder to use in this use case than the `std::vector` API.

Reply:
Answered: morally incorrect.